File: examples/ios/MSDKSampleAppObjC/MSDKSampleAppObjC/HomeViewController.m
//
//  HomeViewController.m
//  MSDKSampleAppObjC
//
//  Created by Norman Jarvis on 5/3/19.
//  Copyright © 2019 Evernym Inc. All rights reserved.
//
#import "CustomTableViewCell.h"
#import "HomeViewController.h"
#import "MobileSDK.h"
#import "Connection.h"
#import "Credential.h"
#import "Utilities.h"
#import "LocalStorage.h"
#import "QRCodeReaderViewController.h"
#import "QRCodeReader.h"
#import "Message.h"
#import "ProofRequest.h"
#import "ConnectionInvitation.h"
#import "ConnectionHandler.h"
#import "CredentialOffersHandler.h"
#import "ProofRequestsHandler.h"

@interface HomeViewController ()

@property (weak, nonatomic) IBOutlet UILabel *infoLbl;
@property (nonatomic, readwrite, strong) IBOutlet UITableView *tableView;
@property (retain, nonatomic) IBOutlet UIButton *addConnectionButton;
@property (retain, nonatomic) IBOutlet UILabel *newConnLabel;

@property NSDictionary* requests;
@property BOOL isInitialized;

@end

@implementation HomeViewController
@synthesize addConnLabel, addConnConfigTextView, requests, newConnLabel, addConnectionButton;

NSString *CREDENTIAL_OFFER = @"credential-offer";
NSString *PRESENTATION_REQUEST = @"presentation-request";
NSString *COMMITTED_QUESTION = @"committed-question";
NSString *OOB = @"OOB";

UIGestureRecognizer *tapper;

- (void)viewDidLoad {
    // On iOS diveces you cann't use qr code scnner
    // But you can handle invite with text input and button on home page
    // For show input set this flag to false
    BOOL const isHideInput = true;
    
    [super viewDidLoad];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    addConnConfigTextView.delegate = self;
    addConnConfigTextView.layer.cornerRadius = 5;
    [addConnConfigTextView setHidden: isHideInput];
    [newConnLabel setHidden: isHideInput];
    [addConnectionButton setHidden:isHideInput];
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesBegan:touches withEvent:event];
    [addConnConfigTextView endEditing:true];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear: animated];
    requests = [LocalStorage getObjectForKey: @"requests" shouldCreate: false];
    [self.tableView reloadData];
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(vcxInitialized) name:@"vcxInitialized" object: nil];
    _isInitialized = [[MobileSDK shared] sdkInited];
}

- (BOOL) textViewShouldBeginEditing:(UITextView *)textView {
    if([textView.text isEqual: @"enter code here"]) {
        addConnConfigTextView.text = @"";
    }
    return true;
}

- (void) viewWillDisappear:(BOOL)animated {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
}

- (void) handleSingleTap:(UITapGestureRecognizer *) sender {
    [self.view endEditing: YES];
}

- (void) vcxInitialized {
    self.infoLbl.text = @"VCX initialized!";
    _isInitialized = true;
    double delayInSeconds = 10.0;
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds * NSEC_PER_SEC));
    dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
        self.infoLbl.text = @"";
    });
}

- (void) createActionWithInvitation:(NSString *) data {
    NSDictionary *connectValues = [ConnectionInvitation parsedInvite: data];
    NSString *label = [connectValues objectForKey: @"label"];
    NSString *goal = @"";
    if ([connectValues valueForKey:@"goal"] != nil) {
        goal = [connectValues objectForKey: @"goal"];
    } else {
        goal = @"New connection";
    }
    NSString *profileUrl = [connectValues objectForKey: @"profileUrl"];
    
    [self createAction:label
            profileUrl:profileUrl
                  goal:goal
                  type:OOB
                  data:[Utilities dictToJsonString:connectValues]
        additionalData:@""
                 pwDid:@""];
}

- (void) createAction:(NSString *) label
           profileUrl:(NSString *) profileUrl
                 goal:(NSString *) goal
                 type:(NSString *) type
                 data:(NSString *) data
       additionalData:(NSString *) additionalData
                pwDid:(NSString *) pwDid {
    NSMutableDictionary* allRequests = [[LocalStorage getObjectForKey: @"requests" shouldCreate: true] mutableCopy];
    NSString *uuid = [[NSUUID UUID] UUIDString];

    NSDictionary* request = @{
        @"name": label,
        @"profileUrl": profileUrl,
        @"goal": goal,
        @"uuid": uuid,
        @"type": type,
        @"data": data,
        @"additionalData": additionalData,
        @"pwDid": pwDid
    };
    
    [allRequests setValue: request forKey: uuid];
    [LocalStorage store: @"requests" andObject: allRequests];
    requests = allRequests;
    [self.tableView reloadData];
}

- (IBAction)addNewConnBtnClick: (id)sender {
    if(addConnConfigTextView.text.length > 3 && ![addConnConfigTextView.text isEqual: @"enter code here"]) {
        [self createActionWithInvitation: addConnConfigTextView.text];
    }
}

- (IBAction)scanQR: (UIButton*) sender {
    if(!_isInitialized) {
        NSLog(@"Please wait for VCX to initialize!");
        return;
    }
    QRCodeReader *reader = [QRCodeReader readerWithMetadataObjectTypes:@[AVMetadataObjectTypeQRCode]];
    QRCodeReaderViewController *vc = [QRCodeReaderViewController readerWithCancelButtonTitle:@""
                                                                                  codeReader:reader
                                                                         startScanningAtLoad:YES
                                                                      showSwitchCameraButton:YES
                                                                             showTorchButton:YES];
    vc.modalPresentationStyle = UIModalPresentationFormSheet;
    vc.delegate = self;

    [reader setCompletionWithBlock:^(NSString *resultAsString) {
        NSLog(@"%@", resultAsString);
        [vc dismissViewControllerAnimated:YES completion:^{
            [self createActionWithInvitation: resultAsString];
        }];
    }];
    
    [self presentViewController: vc animated: YES completion:^{
        NSLog(@"QR code scanner presented");
    }];
}

- (IBAction)checkMessages:(id)sender {
    [Message downloadAllMessages:^(NSArray *responseArray, NSError *error) {
        for (NSInteger i = 0; i < responseArray.count; i++) {
            NSDictionary *message = responseArray[i];

            NSString *type = [message objectForKey:@"type"];
            if ([type isEqual:CREDENTIAL_OFFER]) {
                [self handleReceivedCredentialOffer:message withCompletionBlock:^(BOOL result, NSError *error) {
                    if (error && error.code > 0) {
                        [Utilities printError:error];
                    }
                }];
            }
            if ([type isEqual:PRESENTATION_REQUEST]) {
                [self handleReceivedProofRequest:message
                             withCompletionBlock:^(BOOL result, NSError *error) {
                    if (error && error.code > 0) {
                        [Utilities printError:error];
                    }
                }];
            }
            if ([type isEqual:COMMITTED_QUESTION]) {
                [self handleReceivedQuestion:message];
            }
        }
    }];
}

- (void) handleReceivedCredentialOffer:(NSDictionary *) message
                   withCompletionBlock:(ResponseWithBoolean) completionBlock {
    [self messageStatusUpdate: message];
    NSString *pwDid = [message objectForKey:@"pwDid"];
    NSString *payloadJson = [message objectForKey:@"payload"];
    NSArray *payloadArray = [Utilities jsonToArray:payloadJson];
    NSDictionary *payloadDict = payloadArray[0];
    
    [Credential createWithOffer:payloadJson
          withCompletionHandler:^(NSDictionary *createdOffer, NSError *error) {
        if (error && error.code > 0) {
            return completionBlock(nil, error);
        }
        
        [self createAction:[payloadDict objectForKey:@"claim_name"]
                profileUrl:@""
                      goal:@"Credential offer"
                      type:CREDENTIAL_OFFER
                      data:payloadJson
            additionalData:[Utilities dictToJsonString:createdOffer]
                     pwDid:pwDid];
    }];
}

- (void) handleReceivedProofRequest:(NSDictionary *) message
                withCompletionBlock:(ResponseWithBoolean) completionBlock{
    [self messageStatusUpdate: message];
    NSString *pwDid = [message objectForKey:@"pwDid"];
    NSString *payload = [message objectForKey:@"payload"];
    NSDictionary *payloadDict = [Utilities jsonToDictionary:payload];

    [ProofRequest createWithRequest:payload
              withCompletionHandler:^(NSDictionary *request, NSError *error) {
        if (error && error.code > 0) {
            return completionBlock(nil, error);
        }
        
        [self createAction:[payloadDict objectForKey:@"comment"]
                profileUrl:@""
                      goal:@"Proof Request"
                      type:PRESENTATION_REQUEST
                      data:payload
            additionalData:[Utilities dictToJsonString:request]
                     pwDid:pwDid];
    }];
}

- (void) handleReceivedQuestion:(NSDictionary *) message {
    [self messageStatusUpdate: message];
    NSString *pwDid = [message objectForKey:@"pwDid"];
    NSString *payload = [message objectForKey:@"payload"];
    NSDictionary *payloadDict = [Utilities jsonToDictionary:payload];
    
    [self createAction:[payloadDict objectForKey:@"question_text"]
            profileUrl:@""
                  goal:@"Question"
                  type:COMMITTED_QUESTION
                  data:[Utilities dictToJsonString:message]
        additionalData:payload
                 pwDid:pwDid];
}

- (void) handleAcceptAction:(NSString *) data
                    forType:(NSString *) forType
                      pwDid:(NSString *) pwDid
             additionalData:(NSString *) additionalData
                       name:(NSString *) name
        withCompletionBlock:(ResponseWithBoolean) completionBlock {
    if ([forType isEqual:OOB]) {
        [ConnectionHandler handleConnectionInvitation:data
                                withCompletionHandler:^(NSString *result, NSError *error) {
            return completionBlock(result, error);
        }];
    }
    if ([forType isEqual:CREDENTIAL_OFFER]) {
        [CredentialOffersHandler acceptCredentialOffer:pwDid
                                            attachment:data
                                          createdOffer:[Utilities jsonToDictionary:additionalData]
                                           fromMessage:true
                                 withCompletionHandler:^(NSString *result, NSError *error) {
            return completionBlock(result, error);
        }];
    }
    if ([forType isEqual:PRESENTATION_REQUEST]) {
        [ProofRequestsHandler acceptProofRequest:pwDid
                                      attachment:[Utilities jsonToDictionary:data]
                                         request:[Utilities jsonToDictionary:additionalData]
                                            name:name
                           withCompletionHandler:^(NSString *result, NSError *error) {
            return completionBlock(result, error);
        }];
    }
    if ([forType isEqual:COMMITTED_QUESTION]) {
        [self answer:data
 withCompletionBlock:^(BOOL result, NSError *error) {
            return completionBlock(result, error);
        }];
    }
}

- (void) handleRejectAction:(NSString *) data
                    forType:(NSString *) forType
                      pwDid:(NSString *) pwDid
             additionalData:(NSString *) additionalData
                       name:(NSString *) name
        withCompletionBlock:(ResponseWithBoolean) completionBlock  {
    if ([forType isEqual:OOB]) {
        return completionBlock(true, nil);
    }
    if ([forType isEqual:CREDENTIAL_OFFER]) {
        [CredentialOffersHandler rejectCredentialOffer:pwDid
                                            attachment:data
                                          createdOffer:[Utilities jsonToDictionary:additionalData] withCompletionHandler:^(NSString *result, NSError *error) {
            return completionBlock(result, error);
        }];
    }
    if ([forType isEqual:PRESENTATION_REQUEST]) {
        [ProofRequestsHandler rejectProofRequest:pwDid
                                         request:additionalData
                                            name:name
                           withCompletionHandler:^(NSString *result, NSError *error) {
            return completionBlock(result, error);
        }];
    }
    if ([forType isEqual:COMMITTED_QUESTION]) {
        return completionBlock(true, nil);
    }
}

- (void)answer:(NSString *) data
withCompletionBlock:(ResponseWithBoolean) completionBlock {
    NSDictionary *message = [Utilities jsonToDictionary:data];
    NSString *payload = [message objectForKey:@"payload"];
    NSDictionary *payloadDict = [Utilities jsonToDictionary:payload];
    NSArray *responses = [payloadDict objectForKey:@"valid_responses"];

    NSString *pwDidMes = [message objectForKey:@"pwDid"];
    NSString *questionsConnection = [ConnectionInvitation getConnectionByPwDid:pwDidMes];

    UIAlertController * alert = [UIAlertController
                                     alertControllerWithTitle:[payloadDict objectForKey:@"question_text"]
                                     message:[payloadDict objectForKey:@"question_detail"]
                                     preferredStyle:UIAlertControllerStyleAlert];
    
    for(NSInteger i = 0; i < responses.count; i++) {
        NSDictionary *response = responses[i];
        UIAlertAction* button = [UIAlertAction
                                actionWithTitle:[response objectForKey:@"text"]
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    [Message answerQuestion:questionsConnection
                                                 message:payload
                                                  answer:[Utilities dictToJsonString:response]
                                     withCompletionBlock:^(BOOL result, NSError *error) {
                                        [LocalStorage addEventToHistory:[NSString stringWithFormat:@"%@ - Answer question", [payloadDict objectForKey:@"question_text"]]];
                                        return completionBlock(result, error);
                                    }];
                                }];
        [alert addAction:button];
    }

    [self presentViewController:alert animated:YES completion:nil];
}

- (void) messageStatusUpdate:(NSDictionary *) message {
    dispatch_async(dispatch_get_main_queue(), ^{
        [Message updateMessageStatus:[message objectForKey:@"pwDid"]
                             messageId:[message objectForKey:@"uid"]
                   withCompletionBlock:^(BOOL result, NSError *error) {
            if (error) {
                [Utilities printError:error];
            }
        }];
    });
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [requests.allKeys count];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
   return 110;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    CustomTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    if (cell == nil) {
        cell = [[CustomTableViewCell alloc] init];
    }
    
    NSDictionary *requestDict = requests[requests.allKeys[indexPath.row]];
    NSString *type = [requestDict objectForKey: @"type"];
    NSString *name = [requestDict objectForKey: @"name"];
    NSString *goal = [requestDict objectForKey: @"goal"];
    NSString *uuid = [requestDict objectForKey: @"uuid"];
    NSString *data = [requestDict objectForKey: @"data"];
    NSString *logoUrl = [requestDict objectForKey: @"profileUrl"];
    NSString *pwDid = [requestDict objectForKey: @"pwDid"];
    NSString *additionalData = [requestDict objectForKey: @"additionalData"];

    [cell updateAttribute:name
                 subtitle:goal
                  logoUrl:logoUrl
           acceptCallback:^() {
        [self handleAcceptAction:data
                         forType:type
                           pwDid:pwDid
                  additionalData:additionalData
                            name:name
             withCompletionBlock:^(BOOL result, NSError *error) {
            [self switchActionToHistoryView: uuid];
        }];
                }
            rejectCallback:^() {
        [self handleRejectAction:data
                         forType:type
                           pwDid:pwDid
                  additionalData:additionalData
                            name:name
             withCompletionBlock:^(BOOL result, NSError *error) {
            [self switchActionToHistoryView: uuid];
            if (type == OOB || type == COMMITTED_QUESTION) {
                [self addRejectAction: name];
            }
        }];
            }
     ];
    [cell.accept setTitle:@"Handle" forState:UIControlStateNormal];
    return cell;
}

- (void) switchActionToHistoryView:(NSString *) uuid {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSMutableDictionary* processedDict = nil;
        NSDictionary* req = [LocalStorage getObjectForKey: @"requests" shouldCreate: false];
        for (NSInteger i = 0; i < req.allKeys.count; i++) {
            NSString *key = req.allKeys[i];
            if ([key isEqual: uuid]) {
                [processedDict removeObjectForKey:key];
            }
        }
        [LocalStorage store: @"requests" andObject: processedDict];
        self.requests = processedDict;
        [self.tableView reloadData];
    });
}

- (void) addRejectAction:(NSString *) name {
    [LocalStorage addEventToHistory: [NSString stringWithFormat:@"%@ - Rejected", name]];
    [self.tableView reloadData];
}

@end
     

File: examples/ios/MSDKSampleAppSwift/MSDKSampleAppSwift/HomeViewController.swift
//
//  HomeViewController.swift
//  MSDKSampleAppSwift
//
//  Created by Evernym on 06.07.2021.
//  Copyright © 2021 Norman Jarvis. All rights reserved.
//

import AVFoundation
import UIKit
import QRCodeReader

class HomeViewController: UIViewController, QRCodeReaderViewControllerDelegate, UITextDropDelegate {
    lazy var readerVC: QRCodeReaderViewController = {
        let builder = QRCodeReaderViewControllerBuilder {
            $0.reader = QRCodeReader(metadataObjectTypes: [.qr], captureDevicePosition: .back)
            
            $0.showTorchButton        = true
            $0.showSwitchCameraButton = true
            $0.showCancelButton       = false
            $0.showOverlayView        = true
            $0.rectOfInterest         = CGRect(x: 0.2, y: 0.2, width: 0.6, height: 0.6)
        }
        
        return QRCodeReaderViewController(builder: builder)
    }()
    
    func reader(_ reader: QRCodeReaderViewController, didScanResult result: QRCodeReaderResult) {
      reader.stopScanning()

      dismiss(animated: true, completion: nil)
    }

    func reader(_ reader: QRCodeReaderViewController, didSwitchCamera newCaptureDevice: AVCaptureDeviceInput) {
        print("Switching capture to: \(newCaptureDevice.device)")
    }

    func readerDidCancel(_ reader: QRCodeReaderViewController) {
      reader.stopScanning()

      dismiss(animated: true, completion: nil)
    }
    
    @IBOutlet var tableView: UITableView!
    @IBOutlet var infoLbl: UILabel!
    @IBOutlet var addConnConfigTextView: UITextView!
    @IBOutlet var newConnLabel: UILabel!
    @IBOutlet var addConnectionButton: UIButton!
    
    var requests: [String: Any] = [:]
    var isInitialized = false;
    
    let CREDENTIAL_OFFER = "credential-offer"
    let PRESENTATION_REQUEST = "presentation-request"
    let COMMITTED_QUESTION = "committed-question"
    let OOB = "OOB"
    
    typealias CompletionHandler = (_ result:Bool, _ error:Error?) -> Void

    override func viewDidLoad() {
        // On iOS diveces you cann't use qr code scnner
        // But you can handle invite with text input and button on home page
        // For show input set this flag to false
        let isHideInput = true;
        
        super.viewDidLoad();
        self.tableView.delegate = self;
        self.tableView.dataSource = self;
        self.addConnConfigTextView.delegate = self;
        self.addConnConfigTextView.layer.cornerRadius = 5;
        
        self.addConnConfigTextView.isHidden = isHideInput;
        self.addConnectionButton.isHidden = isHideInput;
        self.newConnLabel.isHidden = isHideInput;
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        self.addConnConfigTextView.endEditing(true)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated);
        if let storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
            self.requests = storageRequests;
            self.tableView.reloadData();
            NotificationCenter.default.addObserver(self, selector: #selector(self.vcxInitialized), name: NSNotification.Name(rawValue: "vcxInitialized"), object: nil)
            isInitialized = MobileSDK.shared().sdkInited
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self)
    }

    @objc func handleSingleTap(_ sender: UITapGestureRecognizer) {
        self.view.endEditing(true)
    }
    
    @objc func vcxInitialized() {
        self.infoLbl.text = "VCX initialized!"
        isInitialized = true;
        let delayInSeconds: Double = 15.0
        DispatchQueue.main.asyncAfter(deadline: .now() + delayInSeconds, execute: {
            self.infoLbl.text = ""
        })
    }
    
    @objc private func createActionWithInvitation(_ data:String) {
        ConnectionInvitation.parsedInvite(data) { connectValues, error in
            let label = (connectValues?["label"] ?? "New connection") as! String
            let goal = (connectValues?["goal"] ?? "New connection") as! String
            let profileUrl = (connectValues?["profileUrl"] ?? "") as! String
            
            self.createAction(
                label,
                profileUrl: profileUrl,
                goal: goal,
                type: self.OOB,
                data: Utilities.toJsonString(connectValues),
                additionalData: "",
                pwDid: ""
            )
        }
    }
    
    private func createAction(_ label: String,
                              profileUrl: String,
                              goal: String,
                              type: String,
                              data: String,
                              additionalData: String,
                              pwDid: String
                              ) {
        if var storageRequests = LocalStorage.getObjectForKey("requests", shouldCreate: false) as? [String: Any] {
            let uuid = UUID().uuidString
            
            let newRequest:[String:Any] = [
                "name": label,
                "goal": goal,
                "profileUrl": profileUrl,
                "uuid": uuid,
                "type": type,
                "data": data,
                "additionalData": additionalData,
                "pwDid": pwDid
            ];
            
            storageRequests[uuid] = newRequest
            LocalStorage.store("requests", andObject: storageRequests)
            
            self.requests = storageRequests
            self.tableView.reloadData()
        }
    }
    
    @IBAction func addNewConnection(_ sender: Any) {
        let connectionText = addConnConfigTextView.text ?? ""
        if connectionText.count > 3 && connectionText != "enter code here" {
            self.createActionWithInvitation(connectionText)
        }
    }
    
    @IBAction func scanQR(_ sender: UIButton) {
        if(!isInitialized) {
            print("Please wait for VCX to initialize!")
            return;
        }
        readerVC.delegate = self
        readerVC.completionBlock = { (result: QRCodeReaderResult?) in
            let invite = (result?.value ?? "") as String
            print(invite)
            self.createActionWithInvitation(invite)
        }
        readerVC.modalPresentationStyle = .formSheet
        present(readerVC, animated: true, completion: nil)
    }
    
    @IBAction func checkMessages(_ sender: Any) {
        Message.downloadAllMessages { messages, error in
            for message in messages ?? [] {
                let msg = message as! [String : String];
                let type = msg["type"];
                
                if type == self.CREDENTIAL_OFFER {
                    self.handleReceivedCredentialOffer(msg, completionHandler: { _, _ in })
                }
                
                if type == self.PRESENTATION_REQUEST {
                    self.handleReceivedProofrequest(msg, completionHandler: { _, _ in })
                }
                
                if type == "committed-question" {
                    self.handleReceivedQuestion(msg)
                }
            }
        }
    }
    
    private func handleReceivedCredentialOffer(_ message: [String : String], completionHandler: @escaping CompletionHandler) {
        self.messageStatusUpdate(message)
        let pwDid = message["pwDid"]  ?? ""
        let payload = message["payload"] ?? ""
        let payloadArr = convertToArray(text:payload)
        let payloadDict = payloadArr?.first
        let name = (payloadDict?["claim_name"] ?? "Offer") as! String
        
        Credential.create(withOffer: payload) { createdOffer, _ in
            self.createAction(
                name,
                profileUrl: "",
                goal: "Credential Offer",
                type: self.CREDENTIAL_OFFER,
                data: payload,
                additionalData: Utilities.dict(toJsonString: createdOffer),
                pwDid: pwDid
            )
            
            return completionHandler(true, nil)
        }
    }
    
    private func handleReceivedProofrequest(_ message: [String : String], completionHandler: @escaping CompletionHandler) {
        self.messageStatusUpdate(message)
        let pwDid = message["pwDid"]  ?? ""
        let payload = message["payload"] ?? ""
        let payloadDict = convertToDictionary(text:payload)
        let name = payloadDict?["comment"] as? String ?? ""
        
        ProofRequest.create(withRequest: payload) { request, _ in
            self.createAction(
                name,
                profileUrl: "",
                goal: "Proof Request",
                type: self.PRESENTATION_REQUEST,
                data: payload,
                additionalData: Utilities.dict(toJsonString: request),
                pwDid: pwDid
            )
            
            return completionHandler(true, nil)
        }
    }
    
    private func handleReceivedQuestion(_ message: [String : String]) {
        self.messageStatusUpdate(message)
        let pwDid = message["pwDid"]  ?? ""
        let payload = message["payload"] ?? ""
        let payloadDict = convertToDictionary(text:payload)
        let name = payloadDict?["question_text"] as? String ?? ""
        self.createAction(
            name,
            profileUrl: "",
            goal: "Question",
            type: self.COMMITTED_QUESTION,
            data: Utilities.toJsonString(message),
            additionalData: payload,
            pwDid: pwDid
        )
    }
    
    @objc private func handleAcceptAction (_ data: String,
                                           forType: String,
                                           pwDid: String,
                                           additionalData: String,
                                           name: String,
                                           completionHandler: @escaping CompletionHandler) -> Any? {
        if forType == self.OOB {
            ConnectionHandler.handleConnectionInvitation(data) { result, error in
                return completionHandler(true, nil)
            }
        }
        if forType == self.CREDENTIAL_OFFER {
            CredentialOffersHandler.acceptCredentialOffer(pwDid,
                                                          attachment: data,
                                                          createdOffer: Utilities.json(toDictionary: additionalData),
                                                          fromMessage: true
            ) { _, _ in
                return completionHandler(true, nil)
            }
        }
        if forType == self.PRESENTATION_REQUEST {
            ProofRequestsHandler.acceptProofRequest(pwDid,
                                                    attachment: Utilities.json(toDictionary: data),
                                                    request: Utilities.json(toDictionary: additionalData),
                                                    name: name
            ) { _, _ in
                return completionHandler(true, nil)
            }
        }
        if forType == self.COMMITTED_QUESTION {
            _ = self.answer(data) { result, error in
                return completionHandler(result, nil)
            }
        }
        return completionHandler(false, nil)
    }
    
    @objc private func handleRejectAction (_ data: String,
                                           forType: String,
                                           pwDid: String,
                                           additionalData: String,
                                           name: String,
                                           completionHandler: @escaping CompletionHandler) -> Any? {
        if forType == self.OOB {
            return completionHandler(true, nil)
        }
        if forType == self.CREDENTIAL_OFFER {
            CredentialOffersHandler.rejectCredentialOffer(pwDid,
                                                          attachment: Utilities.json(toDictionary: data),
                                                          createdOffer: Utilities.json(toDictionary: additionalData)
            ) { _, _ in
                return completionHandler(true, nil)
            }
        }
        if forType == self.PRESENTATION_REQUEST {
            ProofRequestsHandler.rejectProofRequest(pwDid,
                                                    request: additionalData,
                                                    name: name
            ) { _, _ in
                return completionHandler(true, nil)
            }
        }
        if forType == self.COMMITTED_QUESTION {
            return completionHandler(true, nil)
        }
        return completionHandler(false, nil)
    }
    
    @objc private func answer(_ data: String, completionHandler: @escaping CompletionHandler) -> Any? {
        let msg = convertToDictionary(text: data)
        let payload = msg?["payload"];
        let payloadDict = convertToDictionary(text: payload as! String)
        let responses = payloadDict?["valid_responses"] as? [[String : String]];
        
        let pwDidMes = msg?["pwDid"];
        let questionConnection = ConnectionInvitation.getConnectionByPwDid(pwDidMes as? String ?? "");
        let alert = UIAlertController(
            title: payloadDict?["question_text"] as? String,
            message: payloadDict?["question_detail"] as? String,
            preferredStyle: UIAlertController.Style.alert
        );
        
        for response in responses ?? [] {
            alert.addAction(
                UIAlertAction(
                    title: response["text"],
                    style: UIAlertAction.Style.default
                ) {_ in
                    Message.answerQuestion(questionConnection, message: payload as! String, answer: Utilities.dict(toJsonString: response)) { result, error in
                        LocalStorage.addEvent(toHistory: NSString.localizedStringWithFormat("%@ - Answer question", payloadDict?["question_text"] as! CVarArg) as String);
                        return completionHandler(result, error);
                    }
                }
            )
        }
        self.present(alert, animated: true)
        return completionHandler(false, nil);
    }
    
    @objc private func messageStatusUpdate(_ message: [String : String]) {
        Message.updateStatus(message["pwDid"]!, messageId: message["uid"]!) { _,_ in  }
    }
    
    private func convertToDictionary(text: String) -> [String: Any]? {
        if let data = text.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    private func convertToArray(text: String) -> [[String: Any]]? {
        if let data = text.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    private func requestByIndex(_ index: Int) -> Any? {
        let requestsIDs = Array(requests.keys);
        if requestsIDs.count > 0 {
            let requestsID = requestsIDs[index];
            return requests[requestsID];
        }
        return [:]
    }
    
    private func switchRequestToHistoryView(_ uuid: String, completionHandler: @escaping CompletionHandler) -> Any? {
        var storageRequests = requests;
        print("switchRequestToHistoryView", storageRequests, uuid)
        for key in Array(storageRequests.keys) {
            if key == uuid {
                storageRequests.removeValue(forKey: key);
            }
        }
        LocalStorage.store("requests", andObject: storageRequests);
        self.requests = storageRequests;
        self.tableView.reloadData();
        return completionHandler(true, nil);
    }
    
    private func addRejectAction(_ name: String) {
        LocalStorage.addEvent(toHistory: "\(name) - Rejected")
        self.tableView.reloadData();
        return
    }
}

extension HomeViewController: UITextViewDelegate, UITableViewDelegate, UITableViewDataSource {
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 110;
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1;
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return requests.count;
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: CustomTableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! CustomTableViewCell;
        let request = self.requestByIndex(indexPath.row) as? [String: Any] ?? [:];
        
        let type = request["type"] as? String ?? "";
        let name = request["name"] as? String ?? "";
        let goal = request["goal"] as? String ?? "";
        let uuid = request["uuid"] as? String ?? "";
        let data = request["data"] as? String ?? "";
        let logoPath = request["logoUrl"] as? String ?? "";
        let pwDid = request["pwDid"] as? String ?? "";
        let additionalData = request["additionalData"] as? String ?? "";

        cell.updateCellAttributes(title: name, subtitle: goal, logoUrl: logoPath)
        cell.addAceptCallback(acceptCallback: { () -> () in
            _ = self.handleAcceptAction(data,
                                        forType: type,
                                        pwDid: pwDid,
                                        additionalData: additionalData,
                                        name: name
            ) { _, _ in
                _ = self.switchRequestToHistoryView(uuid) { _, _ in }
            }
        })
        cell.addRejectCallback(rejectCallback: { () -> () in
            _ = self.handleRejectAction(data,
                                        forType: type,
                                        pwDid: pwDid,
                                        additionalData: additionalData,
                                        name: name
            ) { _, _ in
                _ = self.switchRequestToHistoryView(uuid) { _, _ in }
                if type == self.OOB || type == self.COMMITTED_QUESTION {
                    self.addRejectAction(name)
                }
            }
            
        });
        
        return cell;
    }
}
